src/coreclr/nativeaot/Runtime/MathHelpers.cpp

```diff
@@ -9,9 +9,29 @@
 // Floating point and 64-bit integer math helpers.
 //
 
+FORCEINLINE int64_t FastDbl2Lng(double val)
+{
+#ifdef TARGET_X86
+    return HCCALL1_V(JIT_Dbl2Lng, val);
+#else
+    return((__int64) val);
+#endif
+}
+
 EXTERN_C NATIVEAOT_API uint64_t REDHAWK_CALLCONV RhpDbl2ULng(double val)
 {
-    return((uint64_t)val);
+    const double two63  = 2147483648.0 * 4294967296.0;
+    uint64_t ret;
+    if (val < two63)
+    {
+        ret = FastDbl2Lng(val);
+    }
+    else
+    {
+        // subtract 0x8000000000000000, do the convert then add it back again
+        ret = FastDbl2Lng(val - two63) + I64(0x8000000000000000);
+    }
+    return ret;
```

Comment:
jit's version seems to have a separate path for `val` being negative: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/utils.cpp#L1972-L1988

Reply:
This implementation is from:

https://github.com/dotnet/runtime/blob/bab1a1f239325746873d2ae795ea343e7901dbce/src/coreclr/vm/jithelpers.cpp#L588-L604